polar-core/src/partial/constraints.rs

```diff
@@ -0,0 +1,501 @@
+    }
+
+    pub fn into_term(self) -> Term {
+        Term::new_temporary(Value::Partial(self))
+    }
+
+    /// Return the expression represented by this partial's constraints.
+    pub fn into_expression(self) -> Term {
+        Term::new_temporary(Value::Expression(Operation {
+            operator: Operator::And,
+            args: self
+                .operations
+                .into_iter()
+                .map(|op| Term::new_temporary(Value::Expression(op)))
+                .collect(),
+        }))
+    }
+
+    pub fn clone_with_name(&self, name: Symbol) -> Self {
+        let mut new = self.clone();
+        new.variable = name;
+        new
+    }
+
+    pub fn clone_with_operations(&self, operations: Vec<Operation>) -> Self {
+        let mut new = self.clone();
+        new.operations = operations;
+        new
+    }
+
+    pub fn name(&self) -> &Symbol {
+        &self.variable
+    }
+
+    fn variable_term(&self) -> Term {
+        Term::new_temporary(Value::Variable(sym!("_this")))
+    }
+}
+
+#[derive(Clone)]
+struct IsaConstraintCheck {
+    existing: Vec<Operation>,
+    proposed_tag: Option<Symbol>,
+    result: Option<bool>,
+    last_call_id: u64,
+}
+
+impl IsaConstraintCheck {
+    pub fn new(existing: Vec<Operation>, mut proposed: Operation) -> Self {
+        let right = proposed.args.pop().unwrap();
+        let proposed_tag = if let Value::Pattern(Pattern::Instance(instance)) = right.value() {
+            Some(instance.tag.clone())
+        } else {
+            None
+        };
+
+        Self {
+            existing,
+            proposed_tag,
+            result: None,
+            last_call_id: 0,
+        }
+    }
+
+    /// Check if constraint is compatible with proposed.
+    ///
+    /// Returns: None if compatible, QueryEvent::Done { false } if incompatible,
+    /// or QueryEvent to ask for compatibility.
+    fn check_constraint(
+        &mut self,
+        mut constraint: Operation,
+        counter: &Counter,
+    ) -> Option<QueryEvent> {
+        if constraint.operator != Operator::Isa {
+            return None;
+        }
+
+        let right = constraint.args.pop().unwrap();
+        if let Value::Pattern(Pattern::Instance(instance)) = right.value() {
+            let call_id = counter.next();
+            self.last_call_id = call_id;
+
+            // is_subclass check of instance tag against proposed
+            return Some(QueryEvent::ExternalIsSubclass {
+                call_id,
+                left_class_tag: self.proposed_tag.clone().unwrap(),
+                right_class_tag: instance.tag.clone(),
+            });
+
+            // TODO check fields for compatibility.
+        }
+
+        None
+    }
+}
+
+impl Runnable for IsaConstraintCheck {
+    fn run(&mut self, counter: Counter) -> PolarResult<QueryEvent> {
+        if self.proposed_tag.is_none() {
+            return Ok(QueryEvent::Done { result: true });
+        }
```

Comment:
When would this happen?

Reply:
If the proposed constraint is not of the form `x matches Type{fields ?}`.  Instead it could be `x matches { }`, which we don't constrain right now.

Reply:
If you try to match a dictionary, not an instance; e.g., `x matches {...}`.